Content.Server/GameObjects/Components/Items/RCDComponent.cs

```diff
@@ -102,50 +105,95 @@ public void Examine(FormattedMessage message, bool inDetailsRange)
+        public void AfterInteract(AfterInteractEventArgs   eventArgs) // TODO: do_after()
         {
             var mapGrid = _mapManager.GetGrid(eventArgs.ClickLocation.GridID);
             var tile = mapGrid.GetTileRef(eventArgs.ClickLocation);
             var coordinates = mapGrid.GridTileToLocal(tile.GridIndices);
             //Less expensive checks first. Failing those ones, we need to check that the tile isn't obstructed.
-            if (_ammo <= 0 || coordinates == GridCoordinates.InvalidGrid || !InteractionChecks.InRangeUnobstructed(eventArgs))
+            if (_ammo <= 0)
+            {
+                _serverNotifyManager.PopupMessage(Owner, eventArgs.User, $"The RCD is out of ammo!");
+                return;
+            }
+            if (coordinates == GridCoordinates.InvalidGrid || !InteractionChecks.InRangeUnobstructed(eventArgs))
             {
                 return;
             }
 
-            var targetTile = (ContentTileDefinition) _tileDefinitionManager[tile.Tile.TypeId];
-
-            var canPlaceTile = targetTile.IsSubFloor; //Boolean to check if we're able to build the desired tile. This defaults to checking for subfloors, but is overridden by "deconstruct" which sets it to the inverse.
+            var snapPos = mapGrid.SnapGridCellFor(eventArgs.ClickLocation, SnapGridOffset.Center);
 
-            switch (this._mode)
+            switch (_mode)
             {
                 //Floor mode just needs the tile to be a space tile (subFloor)
                 case RcdMode.Floors:
+                    if (!tile.Tile.IsEmpty)
+                    {
+                        _serverNotifyManager.PopupMessage(Owner, eventArgs.User, $"You can only build a floor on space!");
+                        return;
+                    }
+
+                    mapGrid.SetTile(eventArgs.ClickLocation, new Tile(_tileDefinitionManager["floor_steel"].TileId));
                     break;
                 //We don't want to place a space tile on something that's already a space tile. Let's do the inverse of the last check.
                 case RcdMode.Deconstruct:
-                    canPlaceTile = !targetTile.IsSubFloor;
+                    if (tile.Tile.IsEmpty)
+                    {
+                        return;
+                    }
+
+                    if (!tile.IsBlockedTurf(true)) //Delete the turf
+                    {
+                        mapGrid.SetTile(snapPos, Tile.Empty);
+                    }
+                    else //Delete what the user targeted
+                    {
+                        //Don't delete mobs or puddles
+                        if (eventArgs.Target == null || eventArgs.Target.TryGetComponent(out PuddleComponent puddleComponent) || eventArgs.Target.TryGetComponent<BodyManagerComponent>(out var bodyManagerComponent))
+                        {
+                            return;
+                        }
+
+                        eventArgs.Target.Delete();
```

Comment:
I don't know, I feel like this should be the opposite: You should only be able to delete entities marked as being deleteable by the RCD

Reply:
Note to self: RCDDeconstructWhitelistComponent